### ISSUE TITLE
BMW712 - current/voltage values are there, not only once

### ISSUE THREAD

**Issue by cbachmann987** [KinDR007/VictronMPPT-ESPHOME#93]
I am trying to integrate my Victron devices into HASS and thanks to this framework I was very successfull with my charger itself, but I don't manage to get measurements out of my BMV 712 Smart BMS - mainly I would be interested in battery voltage, current, SOC from it.

I see the battery current / voltage updated once in the device/entity in HASS, but thats it. On the other hand, I see the log entries passing with the information at hand received by the ESP32 via UART, but somehow they are not updated/send to HASS any more.

What could I possibly have forgotten? As I have the MPPT 150/45 running, I think it is not something substantial.

This is the log ouput from the debug-esp32-example (slightly modified to get it running)

[logs_debug-esp32-example_upload.txt](https://github.com/KinDR007/VictronMPPT-ESPHOME/files/9152851/logs_debug-esp32-example_upload.txt)

I am using a NodeMCU board. I have not much experience with the ESP32 system yet, so please forgive if I overlooked something obvious.

If you need further information, please let me know.

This is the yaml configuration I am trying for my application:

```yaml
substitutions:
  name: "victron-bms"

esphome:
  name: ${name}

esp32:
  board: nodemcu-32s
  framework:
    type: arduino
    version: latest

external_components:
  - source: github://KinDR007/VictronMPPT-ESPHOME@main
    refresh: 0s

wifi:
  ssid: (redacted)
  password: (redacted)

# Enable Home Assistant API
api:
  encryption:
    key: "....."

ota:
  password: "..."


logger:

uart:
  id: uart0
  baud_rate: 19200
  tx_pin: GPIO16
  rx_pin: GPIO17
  debug:
    direction: BOTH
    dummy_receiver: true
    after:
      delimiter: "\n"
    sequence:
      - lambda: UARTDebug::log_string(direction, bytes);

victron:
  uart_id: uart0
  id: victron1
  throttle: 10s

text_sensor:
  - platform: victron
    victron_id: victron1
    model_description:
      name: "${name} model description"
    firmware_version:
      name: "${name} firmware version"
    device_type:
      name: "${name} device type"
    serial_number:
      name: "${name} serial number"
      
      
sensor:
  - platform: victron
    victron_id: victron1
    battery_voltage:
      name: "BMS Battery voltage"
    battery_current:
      name: "BMS Battery current"
```

**Comment by syssi** [comment]
Please remove this line from your configuration:

```
    dummy_receiver: true
```

I assume the dummy receiver steals the victron component the serial data. ;-)

**Comment by syssi** [comment]
It looks like I've messed up the debug examples.

**Comment by cbachmann987** [comment]
> Please remove this line from your configuration:
> 
> ```
>     dummy_receiver: true
> ```
> 
> I assume the dummy receiver steals the victron component the serial data. ;-)

Yes, thanks, stupid me - removed the line and it works. I did not get that this would take all the data from it. I had the complete debug log removed for the other device and it seems that it solved the problem there some time ago for me already without me knowing it.

This helped me a lot, thanks again.

> It looks like I've messed up the debug examples.

Not sure, perhaps its me as well with the changes I made, but I had to select another device/platform to get it running

**Comment by syssi** [comment]
I've prepared a pull request to disable the `dummy_receiver` at the debug examples.

PS. I have a wish: Please double check the sign, decimal place and resolution of all sensors you are using. The BMV support is pretty new.

**Comment by syssi** [comment]
Feel free to close this issue if it's solved.

**Comment by syssi** [comment]
I've checked the response of your BMV and it looks like there is another issue. This is the first device pushing data in (two) chunks:

```
# first chunk
[20:16:18][D][uart_debug:158]: <<< "PID\t0xA381\r\n"
[20:16:18][D][uart_debug:158]: <<< "V\t26457\r\n"
[20:16:19][D][uart_debug:158]: <<< "VS\t34\r\n"
[20:16:19][D][uart_debug:158]: <<< "I\t-6032\r\n"
[20:16:19][D][uart_debug:158]: <<< "P\t-160\r\n"
[20:16:19][D][uart_debug:158]: <<< "CE\t-6673\r\n"
[20:16:19][D][uart_debug:158]: <<< "SOC\t937\r\n"
[20:16:19][D][uart_debug:158]: <<< "TTG\t435\r\n"
[20:16:19][D][uart_debug:158]: <<< "Alarm\tOFF\r\n"
[20:16:19][D][uart_debug:158]: <<< "Relay\tOFF\r\n"
[20:16:19][D][uart_debug:158]: <<< "AR\t0\r\n"
[20:16:19][D][uart_debug:158]: <<< "BMV\t712 Smart\r\n"
[20:16:19][D][uart_debug:158]: <<< "FW\t0408\r\n"
[20:16:19][D][uart_debug:158]: <<< "MON\t0\r\n"
[20:16:19][D][uart_debug:158]: <<< "Checksum\t\x9E"
[20:16:19][D][uart_debug:158]: <<< "\r\n"
```

```
# Second chunk
[20:16:19][D][uart_debug:158]: <<< "H1\t-96470\r\n"
[20:16:19][D][uart_debug:158]: <<< "H2\t-6673\r\n"
[20:16:19][D][uart_debug:158]: <<< "H3\t-26485\r\n"
[20:16:19][D][uart_debug:158]: <<< "H4\t29\r\n"
[20:16:19][D][uart_debug:158]: <<< "H5\t4\r\n"
[20:16:19][D][uart_debug:158]: <<< "H6\t-2316636\r\n"
[20:16:19][D][uart_debug:158]: <<< "H7\t27\r\n"
[20:16:19][D][uart_debug:158]: <<< "H8\t28776\r\n"
[20:16:19][D][uart_debug:158]: <<< "H9\t7135\r\n"
[20:16:19][D][uart_debug:158]: <<< "H10\t31\r\n"
[20:16:19][D][uart_debug:158]: <<< "H11\t0\r\n"
[20:16:19][D][uart_debug:158]: <<< "H12\t0\r\n"
[20:16:19][D][uart_debug:158]: <<< "H15\t22\r\n"
[20:16:19][D][uart_debug:158]: <<< "H16\t38\r\n"
[20:16:19][D][uart_debug:158]: <<< "H17\t6066\r\n"
[20:16:19][D][uart_debug:158]: <<< "H18\t6117\r\n"
[20:16:19][D][uart_debug:158]: <<< "Checksum\tj"
[20:16:19][D][uart_debug:158]: <<< "\r\n"
```

The `throttle` implementation expects one big chunk and uses the "Checksum" as "end of chunk/frame" indicator:

https://github.com/KinDR007/VictronMPPT-ESPHOME/blob/main/components/victron/victron.cpp#L126-L132

It's possible all values of the second chunk are discarded at the moment because of the `throttle` implementation. Which of both chunks will be discarded will be random and depends on "what did arrive first after boot-up". Please disable the trottle feature to be able to receive both chunks:

```
victron:
  uart_id: uart0
  id: victron1
  throttle: 0s
```

**Comment by cbachmann987** [comment]
> I've prepared a pull request to disable the `dummy_receiver` at the debug examples.
ok thanks, that makes sense I think - others may fall into the same trap..

> PS. I have a wish: Please double check the sign, decimal place and resolution of all sensors you are using. The BMV support is pretty new.
I will report problems if I see them, yes.

So as you have opened another issue covering the "divided chunks" topic, I would say we can close here. Thanks for looking at the details of my logs - and if I can be of any help implementing the throttle feature for that as well, please let me know.
